### PR TITLE
🐛 Allow http protocol for noscript > img fallbacks for parity with amp-img

### PR DESCRIPTION
**Pull request by westonruter**
This addresses an issue discovered in the AMP plugin. See https://github.com/ampproject/amp-wp/pull/2051. To repeat that PR's description:

We discovered an issue when a site is served via HTTP instead of HTTPS. An `amp-img` [allows](https://github.com/ampproject/amphtml/blob/7b16d1cc109eb77a8e0477dd8eff9501fae638b2/validator/validator-main.protoascii#L4963-L4979) for its image to use the `http` protocol, so this is valid AMP:

```html
<amp-img src="http://example.com/bison-1024x668.jpg" srcset="http://example.com/bison-1024x668.jpg 1024w, http://example.com/bison-300x196.jpg 300w, http://example.com/bison-768x501.jpg 768w, http://example.com/bison.jpg 1200w" sizes="(max-width: 767px) 89vw, (max-width: 1000px) 54vw, (max-width: 1071px) 543px, 580px" width="1024" height="668" layout="intrinsic"></amp-img>
```

With https://github.com/ampproject/amp-wp/pull/1861 there was introduced `noscript > img` fallbacks for serving when a user with JS disabled accesses a (canonical) AMP page. This results in markup like:

```html
<amp-img src="http://example.com/bison-1024x668.jpg" srcset="http://example.com/bison-1024x668.jpg 1024w, http://example.com/bison-300x196.jpg 300w, http://example.com/bison-768x501.jpg 768w, http://example.com/bison.jpg 1200w" sizes="(max-width: 767px) 89vw, (max-width: 1000px) 54vw, (max-width: 1071px) 543px, 580px" width="1024" height="668" layout="intrinsic">
  <noscript>
    <img src="http://example.com/bison-1024x668.jpg" srcset="http://example.com/bison-1024x668.jpg 1024w, http://example.com/bison-300x196.jpg 300w, http://example.com/bison-768x501.jpg 768w, http://example.com/bison.jpg 1200w" sizes="(max-width: 767px) 89vw, (max-width: 1000px) 54vw, (max-width: 1071px) 543px, 580px" width="1024" height="668">
  </noscript>
</amp-img>
```

It turns out that this actually is invalid AMP:

image

The `src` and `srcset` attributes are flagged as errors because—perhaps unintentionally—the `noscript > img` element [does not allow](https://github.com/ampproject/amphtml/blob/7b16d1cc109eb77a8e0477dd8eff9501fae638b2/validator/validator-main.protoascii#L2235-L2245) the `http` protocol.

Perhaps the `noscript > img` omitted the `http` protocol since it was also omitted for `iframe`, `video`, and `audio`. But `amp-iframe`, `amp-video`, and `amp-audio` all only allow `https` to begin with. The `amp-img` is unique in that it also allows `http`. So this PR ensures that its `noscript > img` fallback element also allows `http`.

### CODE REVIEW

**Comment by twifkak** [comment]
@Gregable: It looks like you explicitly disallowed http URLs on img tags inside noscript, long ago, in commit cee1a3b7 (cl/114347806). Do you know why (or know how to find out)? Does this requirement still hold?

**Comment by Gregable** [comment]
Currently the AMP Cache doesn't rewrite `` URLs (or at least my quick look at the code seems to indicate as such), so if `http` were allowed, we'd end up with mixed-mode warnings (only when javascript is disabled? unsure) which is not desired.

I don't see any reason why we shouldn't rewrite `` URLs to point to the AMP Cache, but we do need to order these changes. I think we never implemented it as it's super unlikely that the amp cache would be loaded without javascript.

**Comment by twifkak** [comment]
Mixed mode warnings would only appear when JS is disabled. I only tested this on Chrome. Also, per the spec:

https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inhead
https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody

> **A start tag whose tag name is "noscript", if the scripting flag is enabled**
> &nbsp;&nbsp;Follow the [generic raw text element parsing algorithm](https://html.spec.whatwg.org/multipage/parsing.html#generic-raw-text-element-parsing-algorithm).

So the contents of the `<noscript>` wouldn't even be tokenized (up until the next `</noscript>`).

**Comment by mrjoro** [comment]
Is this PR still active?

**Comment by westonruter** [comment]
Still something I'd like to see happen. Otherwise we need this workaround which won't work if a site is indeed not available over HTTPS: https://github.com/ampproject/amp-wp/issues/2126

Granted, the frequency of this happening will lessen going into the future, as more sites move to HTTPS and more JS is available on more bots.

**Comment by Gregable** [comment]
The rewrite is now in the AMP cache, from my read of the code, so this should be fine to commit now.

**Comment by mrjoro** [comment]
It looks like this PR was approved; is it still okay to merge it?

**Comment by twifkak** [comment]
@Gregable probably knows best.

File: validator/validator-main.protoascii [hunk 1]
```diff
@@ -2238,6 +2238,7 @@ tags: {
     mandatory: true
     value_url: {
       protocol: "data"
+      protocol: "http" # For parity with amp-img's mandatory-src-or-srcset.
       protocol: "https"
       allow_relative: true  # Will be set to false at a future date.
```
Comment (Gregable) [inline validator/validator-main.protoascii]: Can you remove the comment on this line?

The reason we wanted to set these true was that relative might allow an URL to be http even if it wasn't explicitly mentioned as http. Now that we allow http, it's fine to allow relative.

**Comment by Gregable** [comment]
Spotted one small suggestion and it probably needs to be sync'ed with changes before merging since travis hasn't run for 5 months.

File: validator/validator-main.protoascii (hunk 1 above)
Comment (westonruter) [reply validator/validator-main.protoascii]: Done in 6d60cb73d.

**Comment by westonruter** [comment]
> Spotted one small suggestion and it probably needs to be sync'ed with changes before merging since travis hasn't run for 5 months.

Done! Also merged latest changes from `master`.

**Pull request merged**